pkg/controller/elasticsearch/shutdown/node.go

```diff
@@ -0,0 +1,141 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package shutdown
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	esclient "github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/client"
+	ulog "github.com/elastic/cloud-on-k8s/pkg/utils/log"
+)
+
+var log = ulog.Log.WithName("node-shutdown")
+
+// NodeShutdown implements the shutdown.Interface with the Elasticsearch node shutdown API. It is not safe to call methods
+// on this struct concurrently from multiple go-routines.
+type NodeShutdown struct {
+	c           esclient.Client
+	typ         esclient.ShutdownType
+	reason      string
+	podToNodeID map[string]string
+	shutdowns   map[string]esclient.NodeShutdown
+	once        sync.Once
+}
+
+var _ Interface = &NodeShutdown{}
+
+// NewNodeShutdown creates a new NodeShutdown struct restricted to one type of shutdown (typ); podToNodeID is mapping from
+// K8s Pod name to Elasticsearch node ID; reason is a arbitrary bit of metadata that will be attached to each node shutdown
+// request in Elasticsearch and can help tracking and auditing shutdown requests.
+func NewNodeShutdown(c esclient.Client, podToNodeID map[string]string, typ esclient.ShutdownType, reason string) *NodeShutdown {
+	return &NodeShutdown{
+		c:           c,
+		typ:         typ,
+		podToNodeID: podToNodeID,
+		reason:      reason,
+	}
+}
+
+func (ns *NodeShutdown) initOnce(ctx context.Context) error {
+	var err error
+	ns.once.Do(func() {
+		var r esclient.ShutdownResponse
+		r, err = ns.c.GetShutdown(ctx, nil)
+		if err != nil {
+			err = fmt.Errorf("while getting node shutdowns: %w", err)
+			return
+		}
+		shutdowns := map[string]esclient.NodeShutdown{}
+		for _, ns := range r.Nodes {
+			log.V(1).Info("Existing shutdown", "node", ns.NodeID, "type", ns.Type, "status", ns.Status)
+			shutdowns[ns.NodeID] = ns
+		}
+		ns.shutdowns = shutdowns
+	})
+	return err
+}
+
+func (ns *NodeShutdown) lookupNodeID(podName string) (string, error) {
+	nodeID, exists := ns.podToNodeID[podName]
+	if !exists {
+		return "", fmt.Errorf("node %s currently not member of the cluster", podName)
```

Comment:
This might be expected, for example say that:
1. User creates a cluster
2. User makes a change which prevents the Pod to start, for example by setting a wrong `nodeSelector` in the `podTemplate`
3. Pod is deleted by the operator but the new one is stuck in a `Pending` state
4. User makes a new change in order to fix the configuration
5. The operator attempts to delete the Pod, but since it never started and has not joined the cluster this error is raised:

```
2021-08-20T09:05:47.071+0200    ERROR   manager.eck-operator.controller.elasticsearch-controller        Reconciler error        {"service.version": "1.8.0-SNAPSHOT+mmo", "name": "quickstart", "namespace": "default", "error": "node quickstart-es-data-hot-1 currently not member of the cluster", "errorCauses": [{"error": "node quickstart-es-data-hot-1 currently not member of the cluster"}]}
```

This situation would require the user to manually delete the Pod.

I think it is fine to consider that we can delete a Pod if it has not joined the cluster. We might hit a situation where the Pod has actually started, maybe with a ML job already running on it, but as long as the operator lives in the past I'm not sure there is a great solution to this edge case.

Reply:
That's a really good find, I had not considered this case! I think we have to distinguish between upgrade and downscale here though.

During downscale operations we don't want to ignore this error, even if the Pod is currently not part of the cluster because it cannot be scheduled it might have been at some time and have data worth migrating on a (local) disk? If we just went ahead and deleted the Pod the volume claim would be deleted as well.

During upgrades going ahead with the deletion should be fine as we can assume that the upgrade/configuration change could potentially unblock the situation. 

Do you agree?

Reply:
> During downscale operations we don't want to ignore this error [...]

Good point 👍 

> Do you agree?

Yes !